diff --git a/src/core/ngx_config.h b/src/core/ngx_config.h
index ab73079..a6bbf02 100644
--- a/src/core/ngx_config.h
+++ b/src/core/ngx_config.h
@@ -127,5 +127,6 @@ typedef intptr_t        ngx_flag_t;
 #define NGX_MAX_UINT32_VALUE  (uint32_t) 0xffffffff
 #endif
 
+#define NGX_HTTP_SYSLOG_PATCH
 
 #endif /* _NGX_CONFIG_H_INCLUDED_ */
diff --git a/src/core/ngx_log.c b/src/core/ngx_log.c
index 770a590..97edec2 100644
--- a/src/core/ngx_log.c
+++ b/src/core/ngx_log.c
@@ -139,6 +139,12 @@ ngx_log_error_core(ngx_uint_t level, ngx_log_t *log, ngx_err_t err,
 
     ngx_linefeed(p);
 
+#ifdef NGX_HTTP_SYSLOG_PATCH
+    if (level < NGX_LOG_DEBUG && log->prewrite.handler) {
+        log->prewrite.handler(log->prewrite.data, level - 1, errstr, p - errstr);
+    }
+#endif
+
     (void) ngx_write_fd(log->file->fd, errstr, p - errstr);
 
     if (!ngx_use_stderr
@@ -266,6 +272,10 @@ ngx_log_init(u_char *prefix)
 
     ngx_log.file = &ngx_log_file;
     ngx_log.log_level = NGX_LOG_NOTICE;
+#ifdef NGX_HTTP_SYSLOG_PATCH
+    ngx_log.prewrite.data = NULL;
+    ngx_log.prewrite.handler = NULL;
+#endif
 
     name = (u_char *) NGX_ERROR_LOG_PATH;
 
diff --git a/src/core/ngx_log.h b/src/core/ngx_log.h
index b736aa2..65096e9 100644
--- a/src/core/ngx_log.h
+++ b/src/core/ngx_log.h
@@ -43,6 +43,16 @@
 
 typedef u_char *(*ngx_log_handler_pt) (ngx_log_t *log, u_char *buf, size_t len);
 
+#ifdef NGX_HTTP_SYSLOG_PATCH
+typedef void (*ngx_log_callback_pt) (void *data, ngx_uint_t level, u_char *buf, size_t len);
+
+struct ngx_log_callback_s {
+	ngx_log_callback_pt  handler;
+	void                *data;
+};
+#endif
+
+typedef struct ngx_log_callback_s ngx_log_callback_t;
 
 struct ngx_log_s {
     ngx_uint_t           log_level;
@@ -53,6 +63,10 @@ struct ngx_log_s {
     ngx_log_handler_pt   handler;
     void                *data;
 
+#ifdef NGX_HTTP_SYSLOG_PATCH
+    ngx_log_callback_t   prewrite;
+#endif
+
     /*
      * we declare "action" as "char *" because the actions are usually
      * the static strings and in the "u_char *" case we have to override
diff --git a/src/http/modules/ngx_http_log_module.c b/src/http/modules/ngx_http_log_module.c
index 5152038..f1e796e 100644
--- a/src/http/modules/ngx_http_log_module.c
+++ b/src/http/modules/ngx_http_log_module.c
@@ -8,6 +8,10 @@
 #include <ngx_core.h>
 #include <ngx_http.h>
 
+#ifdef NGX_HTTP_SYSLOG_PATCH
+#include <syslog.h>
+#endif
+
 
 typedef struct ngx_http_log_op_s  ngx_http_log_op_t;
 
@@ -297,6 +301,13 @@ ngx_http_log_handler(ngx_http_request_t *r)
             p = op[i].run(r, p, &op[i]);
         }
 
+#ifdef NGX_HTTP_SYSLOG_PATCH
+        if (r->connection->log->prewrite.handler) {
+            r->connection->log->prewrite.handler(r->connection->log->prewrite.data,
+                LOG_INFO, line, p - line);
+        }
+#endif
+
         ngx_linefeed(p);
 
         ngx_http_log_write(r, &log[l], line, p - line);
